### ISSUE TITLE
Better error message when attempting to instantiate tuple structs with private fields

### ISSUE THREAD

**Issue by estebank** [rust-lang/rust#58017]
[Given](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d7253040c69f885b3a45c3b95e204ad7)

```
mod a {
    pub struct Error(usize);
}

fn main(){
    let x = a::Error(3);
}
```

we emit

```
error[E0603]: tuple struct `Error` is private
 --> src/main.rs:6:16
  |
6 |     let x = a::Error(3);
  |                ^^^^^
```

which is confusing and inconsistent with the same error in a regular struct:

```
error[E0451]: field `a` of struct `a::Error` is private
 --> src/main.rs:6:23
  |
6 |     let x = a::Error {a: 3};
  |                       ^^^^ field `a` is private
```

This has [confused users](https://users.rust-lang.org/t/how-should-this-error-message-be-improved/22138/2) in the past. CC #57951 and #52144, which are somewhat related.

**Comment by csmoe** [comment]
in addition to @Centril 's proposal, it should locate the exact position of the private fields:
```rust
struct Foo(pub i32, i32, pub i32, i32);
                    ^^^           ^^^
```

**Comment by lnicola** [comment]
Is there a good reason why the three test cases give completely different error messages?

**Comment by estebank** [comment]
This is actually a regression from 1.16:

```
error[E0450]: cannot invoke tuple struct constructor with private fields
 --> <source>:6:13
  |
2 |     pub struct Error(usize);
  |                      ------ private field declared here
...
6 |     let x = a::Error(3);
  |             ^^^^^^^^ cannot construct with a private field
```

**Comment by cramertj** [comment]
ping on this, which still reproduces.

**Comment by estebank** [comment]
cc @petrochenkov as this overlaps with a lot of the work he's been doing.

**Comment by pnkfelix** [comment]
triage: I personally don't think this should be P-high, as in I don't think it should be part of the triage docket for T-compiler every week.

But I won't downgrade it yet.

First, lets tag it as E-needs-mentor, and see if anyone steps up to provide instructions for a solution here.

**Comment by da-x** [comment]
Posted a PR to fix this - #65153.

**Comment by jonas-schievink** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/39703
